File: hora/horoscope/dhasa/raasi/chakra.py
""" Chakra Dhasa """
"""
    Antardasa/Bhukthi Lords do not match with JHora - as it is not clear how it is implemented there
    So we start with mahadasa lord as the bhukthi lord.
"""
from hora.panchanga import drik
from hora import utils, const
_dhasa_duration = 10.0; _bhukthi_duration = _dhasa_duration/12.0
def _dhasa_seed(jd,place,lagna_house,lagna_lord_house):
    previous_day_sunset_time = drik.sunset(jd-1, place)[0]
    today_sunset_time = drik.sunset(jd, place)[0]
    today_sunrise_time = drik.sunrise(jd, place)[0]
    tomorrow_sunrise_time = 24.0+drik.sunrise(jd+1, place)[0]
    _,_,_,birth_time = utils.jd_to_gregorian(jd)
    df = abs(today_sunset_time - today_sunrise_time)/6.0
    nf1 = abs(today_sunrise_time-previous_day_sunset_time)/6.0
    nf2 = abs(tomorrow_sunrise_time-today_sunset_time)/6.0
    #print('df',df,'nf1',nf1,'nf2',nf2)
    dawn_start = today_sunrise_time-nf1; dawn_end=today_sunrise_time+nf1
    #print('dawn',dawn_start,dawn_end)
    day_start = dawn_end; day_end = today_sunset_time-nf1
    #print('day',day_start,day_end)
    dusk_start = day_end ; dusk_end = today_sunset_time+nf2
    #print('dusk',dusk_start,dusk_end)
    yday_night_start = -(previous_day_sunset_time+nf1); yday_night_end = today_sunrise_time-nf1
    tonight_start = today_sunset_time+nf2; tonight_end = tomorrow_sunrise_time-nf2
    #print('Night-Yday',yday_night_start,yday_night_end,'Night-today',tonight_start,tonight_end)
    # Night is before dawn_start and after dusk_end
    if birth_time > dawn_start and birth_time < dawn_end: # dawn
        kaala_period = 'Dawn'
        _dhasa_seed = (lagna_house+1)%12
    elif birth_time > dusk_start and birth_time < dusk_end: # dusk
        kaala_period = 'Dusk'
        _dhasa_seed = (lagna_house+1)%12
    elif birth_time > day_start and birth_time < day_end: # Day
        kaala_period = 'Day'
        _dhasa_seed = lagna_lord_house
    elif birth_time > yday_night_start and birth_time < yday_night_end: # yday-night
        kaala_period = 'YDay-Night'
        _dhasa_seed = lagna_house
    elif birth_time > tonight_start and birth_time < tonight_end: # yday-night
        kaala_period = 'ToNight'
        _dhasa_seed = lagna_house    
    return _dhasa_seed
def get_dhasa_antardhasa(dob,tob,place,divisional_chart_factor=1,years=1,months=1,sixty_hours=1,include_antardhasa=False):
    jd_at_dob = utils.julian_day_number(dob, tob)
    jd_years = drik.next_solar_date(jd_at_dob, place, years=years, months=months,sixty_hours=months)
    from hora.horoscope.chart import charts, house
    pp = charts.divisional_chart(jd_years, place, divisional_chart_factor=divisional_chart_factor)
    lagna_house = pp[0][1][0];lagna_lord = house.house_owner_from_planet_positions(pp, lagna_house)
    lagna_lord_house = pp[lagna_lord+1][1][0]
    dhasa_seed = _dhasa_seed(jd_years, place,lagna_house,lagna_lord_house)
    dhasa_info = []
    start_jd = jd_years
    for dhasa_lord in [(dhasa_seed+h)%12 for h in range(12)]:
        if include_antardhasa:
            for bhukthi_lord in [(dhasa_lord+h)%12 for h in range(12)]:
                y,m,d,h = utils.jd_to_gregorian(start_jd)
                dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
                dhasa_info.append((dhasa_lord,bhukthi_lord,dhasa_start,round(_bhukthi_duration,2)))
                start_jd += _bhukthi_duration * const.sidereal_year
        else:
            y,m,d,h = utils.jd_to_gregorian(start_jd)
            dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
            dhasa_info.append((dhasa_lord,dhasa_start,round(_dhasa_duration,2)))
            start_jd += _dhasa_duration * const.sidereal_year
    return dhasa_info
if __name__ == "__main__":
    from hora.tests import pvr_tests
    pvr_tests.chakra_test()

File: pyjhora-web/src/core/dhasa/raasi/chakra.ts
/**
 * Chakra Dasha System
 * Ported from PyJHora chakra.py
 * 
 * Fixed 10-year duration cycles
 * Seed based on time of birth (dawn/day/dusk/night)
 */

import { RASI_NAMES_EN, SIDEREAL_YEAR } from '../../constants';
import { PlanetPosition, getDivisionalChart } from '../../horoscope/charts';
import { getHouseOwnerFromPlanetPositions } from '../../horoscope/house';
import { getPlanetLongitude } from '../../panchanga/drik';
import { sunrise, sunset } from '../../ephemeris/swe-adapter';
import type { Place } from '../../types';
import { julianDayToGregorian } from '../../utils/julian';

// ============================================================================
// TYPES
// ============================================================================

export interface ChakraDashaPeriod {
  rasi: number;
  rasiName: string;
  startJd: number;
  startDate: string;
  durationYears: number;
}

export interface ChakraBhuktiPeriod {
  dashaRasi: number;
  bhuktiRasi: number;
  bhuktiRasiName: string;
  startJd: number;
  startDate: string;
  durationYears: number;
}

export interface ChakraResult {
  mahadashas: ChakraDashaPeriod[];
  bhuktis?: ChakraBhuktiPeriod[];
}

// ============================================================================
// CONSTANTS
// ============================================================================

const YEAR_DURATION = SIDEREAL_YEAR;
const DHASA_DURATION = 10; // Fixed 10 years per sign

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

function getPlanetPositionsArray(jd: number, place: Place, divisionalChartFactor: number): PlanetPosition[] {
  const d1Positions: PlanetPosition[] = [];
  
  for (let planet = 0; planet <= 8; planet++) {
    const longitude = getPlanetLongitude(jd, place, planet);
    d1Positions.push({
      planet,
      rasi: Math.floor(longitude / 30),
      longitude: longitude % 30
    });
  }
  
  if (divisionalChartFactor > 1) {
    return getDivisionalChart(d1Positions, divisionalChartFactor);
  }
  
  return d1Positions;
}

function formatJdAsDate(jd: number): string {
  const { date, time } = julianDayToGregorian(jd);
  const pad = (n: number) => Math.abs(n).toString().padStart(2, '0');
  const hour12 = time.hour % 12 || 12;
  const ampm = time.hour < 12 ? 'AM' : 'PM';
  const yearStr = date.year < 0 ? `${Math.abs(date.year)} BC` : date.year.toString();
  return `${yearStr}-${pad(date.month)}-${pad(date.day)} ${pad(hour12)}:${pad(time.minute)}:${pad(time.second)} ${ampm}`;
}

/**
 * Determine dasha seed based on time of birth relative to dawn/day/dusk/night.
 * Ports Python chakra._dhasa_seed().
 *
 * - Dawn/Dusk: seed = (lagnaHouse + 1) % 12
 * - Day: seed = lagnaLordHouse
 * - Night: seed = lagnaHouse
 */
function getDhasaSeed(
  jd: number,
  place: Place,
  lagnaHouse: number,
  lagnaLordHouse: number
): number {
  const previousDaySunsetTime = sunset(jd - 1, place).localTime;
  const todaySunsetTime = sunset(jd, place).localTime;
  const todaySunriseTime = sunrise(jd, place).localTime;
  const tomorrowSunriseTime = 24.0 + sunrise(jd + 1, place).localTime;

  const { time } = julianDayToGregorian(jd);
  const birthTime = time.hour + time.minute / 60 + time.second / 3600;

  const nf1 = Math.abs(todaySunriseTime - previousDaySunsetTime) / 6.0;
  const nf2 = Math.abs(tomorrowSunriseTime - todaySunsetTime) / 6.0;

  const dawnStart = todaySunriseTime - nf1;
  const dawnEnd = todaySunriseTime + nf1;
  const dayStart = dawnEnd;
  const dayEnd = todaySunsetTime - nf1;
  const duskStart = dayEnd;
  const duskEnd = todaySunsetTime + nf2;
  const ydayNightStart = -(previousDaySunsetTime + nf1);
  const ydayNightEnd = todaySunriseTime - nf1;
  const tonightStart = todaySunsetTime + nf2;
  const tonightEnd = tomorrowSunriseTime - nf2;

  if (birthTime > dawnStart && birthTime < dawnEnd) {
    // Dawn
    return (lagnaHouse + 1) % 12;
  } else if (birthTime > duskStart && birthTime < duskEnd) {
    // Dusk
    return (lagnaHouse + 1) % 12;
  } else if (birthTime > dayStart && birthTime < dayEnd) {
    // Day
    return lagnaLordHouse;
  } else if (birthTime > ydayNightStart && birthTime < ydayNightEnd) {
    // Yesterday night
    return lagnaHouse;
  } else if (birthTime > tonightStart && birthTime < tonightEnd) {
    // Tonight
    return lagnaHouse;
  }

  // Fallback: use lagna house
  return lagnaHouse;
}

// ============================================================================
// MAIN FUNCTION
// ============================================================================

/**
 * Get Chakra Dasha periods
 * Fixed 10-year duration per sign
 */
export function getChakraDashaBhukti(
  jd: number,
  place: Place,
  options: {
    divisionalChartFactor?: number;
    includeBhuktis?: boolean;
  } = {}
): ChakraResult {
  const {
    divisionalChartFactor = 1,
    includeBhuktis = true
  } = options;
  
  const planetPositions = getPlanetPositionsArray(jd, place, divisionalChartFactor);

  // Get lagna house (Sun as proxy) and lagna lord's house
  const lagnaHouse = planetPositions[0]?.rasi ?? 0;
  const lagnaLord = getHouseOwnerFromPlanetPositions(planetPositions, lagnaHouse, false);
  const lagnaLordHouse = planetPositions.find(p => p.planet === lagnaLord)?.rasi ?? 0;

  const dhasaSeed = getDhasaSeed(jd, place, lagnaHouse, lagnaLordHouse);
  
  // Build progression from seed
  const dhasaLords = Array.from({ length: 12 }, (_, h) => (dhasaSeed + h) % 12);
  
  const mahadashas: ChakraDashaPeriod[] = [];
  const bhuktis: ChakraBhuktiPeriod[] = [];
  let startJd = jd;
  
  for (const dhasaLord of dhasaLords) {
    const rasiName = RASI_NAMES_EN[dhasaLord] ?? `Rasi ${dhasaLord}`;
    
    mahadashas.push({
      rasi: dhasaLord,
      rasiName,
      startJd,
      startDate: formatJdAsDate(startJd),
      durationYears: DHASA_DURATION
    });
    
    if (includeBhuktis) {
      const bhuktiDuration = DHASA_DURATION / 12;
      let bhuktiStartJd = startJd;
      
      for (let h = 0; h < 12; h++) {
        const bhuktiLord = (dhasaLord + h) % 12;
        const bhuktiRasiName = RASI_NAMES_EN[bhuktiLord] ?? `Rasi ${bhuktiLord}`;
        
        bhuktis.push({
          dashaRasi: dhasaLord,
          bhuktiRasi: bhuktiLord,
          bhuktiRasiName,
          startJd: bhuktiStartJd,
          startDate: formatJdAsDate(bhuktiStartJd),
          durationYears: bhuktiDuration
        });
        
        bhuktiStartJd += bhuktiDuration * YEAR_DURATION;
      }
    }
    
    startJd += DHASA_DURATION * YEAR_DURATION;
  }
  
  return includeBhuktis ? { mahadashas, bhuktis } : { mahadashas };
}
